### PR TITLE
Monitor TBTC token total supply

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Closes: https://github.com/keep-network/tbtc-v2/issues/552

Here we add a new supply monitor that produces a new kind of system event:

### High TBTC token total supply change

A **critical system event** indicating that a high change (i.e. >=10%) of the total TBTC v2 token supply took place in the last 12 hours. This event is sent to Sentry hub and requires an immediate team’s action. The default action is checking the root cause of the supply change and making sure its source is actually a proper deposit/redemption and there are no signs of any malicious action.

### CODE REVIEW

**Comment by pdyraga** [comment]
How this code was tested?

**Comment by lukasz-zimnoch** [comment]
> How was this code tested?

Similarly, to https://github.com/keep-network/tbtc-v2/pull/556. Manual block window on points where the supply changed enough to cause an alert.

File: monitoring/src/supply-monitor.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    // eslint-disable-next-line no-console
+    console.log("running supply monitor check")
+
+    // By default, the supply change is checked between the current and
+    // reference blocks.
+    const currentBlock = toBlock
+    let referenceBlock =
+      currentBlock - totalSupplyBlockSpan > 0
+        ? currentBlock - totalSupplyBlockSpan
+        : 0
```
Comment (pdyraga) [inline monitoring/src/supply-monitor.ts:55]: I am confused about the reference block. If the time span between `fromBlock` and `toBlock` is less than 12h on average, are we checking between `fromBlock` and... `0`?

Could be worth dropping a comment about this case, ideally explaining it on a timeline.

Comment (lukasz-zimnoch) [reply monitoring/src/supply-monitor.ts:55]: > If the time span between fromBlock and toBlock is less than 12h on average, are we checking between fromBlock and... 0?

Negative. This is just a sanity check to not nuke the monitoring if the reference block is <0. If the time span is less than 12h, the reference block is replaced by the last block the alert was fired on.

Comment (lukasz-zimnoch) [reply monitoring/src/supply-monitor.ts:55]: Left an additional explanation: 567dd2dd

File: monitoring/src/supply-monitor.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    change: `${difference.gte(0) ? "+" : "-"}${change.toString()}%`,
+    threshold: `${threshold}%`,
+    referenceBlock: referenceBlock.toString(),
+  },
+  block: currentBlock,
+})
+
+export interface SupplyMonitorPersistence {
+  lastHighTotalSupplyChangeBlock: () => Promise<number>
+  updateLastHighTotalSupplyChangeBlock: (block: number) => Promise<void>
```
Comment (pdyraga) [inline monitoring/src/supply-monitor.ts:32]: Could be worth documenting those two functions in two sentences. 

My understanding of `lastHighTotalSupplyChangeBlock` is that it represents the end block of the check when the alert of the high total supply change` was triggered the last time. If the alert was never triggered, this is 0.

Is this correct?

File: monitoring/src/supply-monitor.ts (hunk 1 above)
Comment (pdyraga) [reply monitoring/src/supply-monitor.ts:55]: I think I see it now. For this particular monitor, `fromBlock` is ignored and the `referenceBlock` replaces it. 

Can we add a small comment on the top of `referenceBlock` to explain it?

File: monitoring/src/supply-monitor.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+      threshold = totalSupplyChangeThreshold / 2
+    }
+
+    const referenceSupply = await this.tbtcToken.totalSupply(referenceBlock)
+    const currentSupply = await this.tbtcToken.totalSupply(currentBlock)
+
+    const difference = currentSupply.sub(referenceSupply)
+    // There is a precision loss here, but it is acceptable as long as the
+    // threshold is at least 1%.
+    const change = difference.abs().mul(100).div(referenceSupply)
```
Comment (pdyraga) [inline monitoring/src/supply-monitor.ts:87]: It is nice it works for + and - case!

File: monitoring/src/supply-monitor.ts (hunk 2 above)
Comment (lukasz-zimnoch) [reply monitoring/src/supply-monitor.ts]: Yep. Added a comment faea9554

File: monitoring/src/supply-monitor.ts (hunk 1 above)
Comment (lukasz-zimnoch) [reply monitoring/src/supply-monitor.ts:63]: faea9554

**Review by pdyraga — APPROVED** [review]

**Pull request merged**